field_utils/tests/Test_FieldCondensedArray_private.pf

```diff
@@ -0,0 +1,147 @@
+       actual = get_array_shape(gridToFieldMap, localElementCount, vertical_dims)
+       @assertEqual(actual, expected, GENERIC_MESSAGE)
+
+    end subroutine test_get_array_shape_general
+
+    @Test
+    subroutine test_get_array_shape_noz()
+       integer :: expected(3), actual(3)
+       integer, allocatable :: gridToFieldMap(:)
+       integer, allocatable :: localElementCount(:)
+
+       gridToFieldMap = [1, 2]
+       localElementCount = [2, 3, 5, 7]
+       expected = [product(localElementCount(1:2)), 1, product(localElementCount(3:))]
+       actual = get_array_shape(gridToFieldMap, localElementCount)
+       @assertEqual(actual, expected, GENERIC_MESSAGE)
+
+    end subroutine test_get_array_shape_noz
+
+    @Test
+    subroutine test_get_array_shape_0D()
+       integer :: expected(3), actual(3)
+       integer, allocatable :: gridToFieldMap(:)
+       integer, allocatable :: localElementCount(:)
+
+       gridToFieldMap = [0, 0]
+       localElementCount = [5, 7, 11]
+       expected = [1, 1, product(localElementCount)]
+       actual = get_array_shape(gridToFieldMap, localElementCount)
+       @assertEqual(actual, expected, GENERIC_MESSAGE)
+
+    end subroutine test_get_array_shape_0D
+
+    subroutine test_get_array_shape_vert_only()
+       integer :: expected(3), actual(3)
+       integer, allocatable :: gridToFieldMap(:)
+       integer, allocatable :: localElementCount(:)
+       integer, allocatable :: vertical_dims(:)
+
+       gridToFieldMap = [0, 0]
+       vertical_dims = [3]
+       localElementCount = vertical_dims
+       expected = [1, localElementCount(1), 1]
+       actual = get_array_shape(gridToFieldMap, localElementCount, vertical_dims)
+       @assertEqual(actual, expected, GENERIC_MESSAGE)
+
+    end subroutine test_get_array_shape_vert_only
+
+    subroutine test_get_array_shape_vert_ungrid()
+       integer :: expected(3), actual(3)
+       integer, allocatable :: gridToFieldMap(:)
+       integer, allocatable :: localElementCount(:)
+       integer, allocatable :: vertical_dims(:)
+
+       gridToFieldMap = [0, 0]
+       vertical_dims = [3]
+       localElementCount = [vertical_dims, 5, 7]
+       expected = [1, localElementCount(1), product(localElementCount(2:))]
+       actual = get_array_shape(gridToFieldMap, localElementCount, vertical_dims)
+       @assertEqual(actual, expected, GENERIC_MESSAGE)
+
+    end subroutine test_get_array_shape_vert_ungrid
+    
+    @Test
+    subroutine test_get_array_shape_2D_ungrid()
+       integer :: expected(3), actual(3)
+       integer, allocatable :: gridToFieldMap(:)
+       integer, allocatable :: localElementCount(:)
+
+       gridToFieldMap = [1, 2]
+       localElementCount = [3, 5, 7, 11]
+       expected = [product(localElementCount(1:2)), 1, product(localElementCount(3:))]
+       actual = get_array_shape(gridToFieldMap, localElementCount)
+       @assertEqual(actual, expected, GENERIC_MESSAGE)
```

Comment:
`assert_that` will provide much better error messages for these.

Reply:
Or at least better.

Reply:
What is the syntax for that? I didn't find that in the pFUnit documentation I have. Is that recent?

Reply:
Not new, but probably/possibly newer than documentation.  There is an entire "hamcrest" layer that provides a much more flexible manner of composing messages and asserting combinations of things.  Unfortunately, it is largely limited to working on Fortran intrinsic types due to lack of support for multiple inheritance.   Forces users to make their type extend a pFUnit base type which is generally ill advised.

Reply:
Typical usage is:
```
@assert_that(found, is(expected))
```
or
```
@assert_that(message, found, is(expected))
```
or even the more verbose:
```
@assert_that(found, is(equal_to(expected)))
```
And there are `less_then` and variants.  As well as `nealy_equal_to` which takes a tolerance.  And advanced stuff like `all` and `any`.
```
And then the caveats.  Don't use it with strings - gfortran is broken.